client/my-sites/comments/comment/comment-html-editor.jsx

```diff
@@ -0,0 +1,242 @@
+/** @format */
+/**
+ * External dependencies
+ */
+import React, { Component } from 'react';
+import PropTypes from 'prop-types';
+import { localize } from 'i18n-calypso';
+import classNames from 'classnames';
+import { delay, each, filter, get, map, reduce } from 'lodash';
+
+/**
+ * Internal dependencies
+ */
+import AddImageDialog from 'post-editor/editor-html-toolbar/add-image-dialog';
+import AddLinkDialog from 'post-editor/editor-html-toolbar/add-link-dialog';
+import Button from 'components/button';
+
+export class CommentHtmlEditor extends Component {
+	static propTypes = {
+		commentContent: PropTypes.string,
+		disabled: PropTypes.bool,
+		onChange: PropTypes.func,
+	};
+
+	state = {
+		openTags: [],
+		selectedText: '',
+		showImageDialog: false,
+		showLinkDialog: false,
+	};
+
+	storeTextareaRef = textarea => ( this.textarea = textarea );
+
+	isTagOpen = tag => -1 !== this.state.openTags.indexOf( tag );
+
+	setCursorPosition = ( selectionEnd, insertedContentLength ) => {
+		this.textarea.selectionEnd = this.textarea.selectionStart =
+			selectionEnd + insertedContentLength;
+	};
+
+	splitSelectedContent = () => {
+		const { selectionEnd, selectionStart, value } = this.textarea;
+		return {
+			before: value.substring( 0, selectionStart ),
+			inner: value.substring( selectionStart, selectionEnd ),
+			after: value.substring( selectionEnd, value.length ),
+		};
+	};
+
+	insertHtmlTag = (
+		tag,
+		attributes,
+		options = {
+			adjustCursorPosition: 0,
+			alsoClose: false,
+			indent: false,
+			newLineAfter: false,
+			paragraph: false,
+			selfClosed: false,
+			text: null,
+		}
+	) => {
+		const element = document.createElement( tag );
+		each( attributes, ( value, key ) => element.setAttribute( key, value ) );
+		element.innerHTML = '<!---->';
+		const fragments = element.outerHTML.split( '<!---->' );
+		const opener =
+			( options.paragraph ? '\n' : '' ) +
+			( options.indent ? '\t' : '' ) +
+			fragments[ 0 ] +
+			( options.paragraph ? '\n' : '' );
+		const closer =
+			fragments[ 1 ] + ( options.newLineAfter ? '\n' : '' ) + ( options.paragraph ? '\n\n' : '' );
+		const inner = options.text || this.splitSelectedContent().inner;
+
+		if ( inner.length || options.alsoClose ) {
+			return this.insertContent( opener + inner + closer, options.adjustCursorPosition );
+		}
+
+		if ( ! options.selfClosed && this.isTagOpen( tag ) ) {
+			this.setState( ( { openTags } ) => ( { openTags: filter( openTags, tag ) } ) );
```

Comment:
careful! we're working with a tree data structure (HTML/DOM) not a list

```js
<ul><li>outer<ul><li>inner</li> // oops wiped out the outer `li` too
```

Reply:
I don't think I got what you mean.

`state.openTags` is a simple array that gets populated whenever you open a tag **via the toolbar**.
e.g. you click on the [B] button, and `openTags.concat( 'b' )`

`openTags` then is used to naively check if a button should open or close a tag.
e.g. `<button>{ this.isTagOpen( 'b' ) ? '/b' : 'b' }</button>`

when you close the tag (again: only **via the toolbar**), we simply remove it from the list.
e.g `filter( openTags, tag => 'b' === tag )`

Then, if there are open tags, the [CLOSE TAGS] button is enabled, and on click it simply insert the close tag for every element in the list.

If we open a tag via the toolbar and then we close it manually by writing `</tag>`, `state.openTags` goes "out of sync" with the content.

So, in other words, `state.openTags` doesn't do anything to the text.

---

This is exactly the same behaviour of the post editor HTML toolbar, and AFAIK also of wp-admin HTML toolbars.

If you have a better approach (and I'm sure there are, but all quite complicated from a RegExp standpoint), I'd suggest to ship this PR and improve after.

If you're pointing out a bug, could you please explain it more? 😄

Reply:
ah. in other words, it's not our open/close code that's broken, it's the toolbar which doesn't express the capabilities of HTML…

in that case also check out `lodash.union` and `lodash.difference` for a nice way to add some semantics to the operations of adding and removing an item from a set. there's a native `Set()` in JavaScript built for this but it's not immutable sadly

Reply:
Well, we're not creating a WYSIWYG editor, but really just a text editor with some helper buttons.
I've tried with some regex magic to keep track of the open/closed tags, but my skills (Google-foo included) are not nearly enough.

I guess I could try to create some sort of tree structure that mirrors the comment content DOM, but I'd say that's hella overkill for this purpose.

I like the idea of using `Set` though!

Reply:
well if we immediately switch the button to the tag closer then we're bound by the UI. I think it's okay though. like you said, they are just helpers.

RegExps will fail us - they are guaranteed to.

Reply:
So in a follow up I'd recommend using [DomParser](https://developer.mozilla.org/en-US/docs/Web/API/DOMParser) to do this for us if it's available:

![screen shot 2018-02-15 at 10 58 53 am](https://user-images.githubusercontent.com/1270189/36275318-57d545f2-123f-11e8-8053-290cd856bc17.png)

Reply:
Thanks @gwwar - that was more helpful than my response. `DOMParser` should be available in every environment we support.